### ISSUE TITLE
Boxing problem appearing on augmented `List#join` when all tests ran

### ISSUE THREAD

**Issue by danielpetisme** [eclipse/golo-lang#268]
Teaser: This is a tricky one

In the standard-augmentations, the current impl of `List#join` (mine in fact :confused:) raise an exception when all tests are ran but not if your run the failing-ish tests.

*how to reproduce*
Add the following method in any golo code used for tests
```
function plop = -> list[1, 2, 3]: join("-")
```
Add the associated Java code
```
Method plop = moduleClass.getMethod("plop");
assertThat((String) plop.invoke(null), is("1-2-3"));
```

Now run `mvn clean test`, 2 tests fail:
* `DocumentationRenderingTest.html_processor`
* `DocumentationRenderingTest.markdown_processor`
```
java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Number
        at sun.invoke.util.ValueConversions.primitiveConversion(ValueConversions.java:242)
        at sun.invoke.util.ValueConversions.unboxInteger(ValueConversions.java:76)
        at gololang.StandardAugmentations$java$util$List.join(standard-augmentations.golo:430)
        at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:625)
        at fr.insalyon.citi.golo.runtime.MethodInvocationSupport.fallback(MethodInvocationSupport.java:191)
        at anonymous317906307738923.__$$_closure_0($Anonymous$_317906307759624.golo:137)
        at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:625)
        at fr.insalyon.citi.golo.doc.MarkdownProcessor.render(MarkdownProcessor.java:34)
        at fr.insalyon.citi.golo.doc.DocumentationRenderingTest.markdown_processor(DocumentationRenderingTest.java:53)
```
if you run unitary the test `mvn -Dtest=DocumentationRenderingTest#html_processor clean test` it passes.

Fix:
In fact I know how to fix the issue. It comes from the current `join` implementation
```
1  function join = |this, separator| {
2     var buffer = java.lang.StringBuilder("")
3     if not (this: isEmpty()) {
4       buffer: append(this: head())
5       let tail = this: tail()
6       if not (tail: isEmpty()) {
7         buffer: append(separator)
8         buffer: append(tail: join(separator))
9       }
10    }
11   return buffer: toString()
12  }
```

Replacing the line`#4` by `buffer: append(this: head(): toString())` (adding a `toString())` fix the issue.

Questions:
1- `toString` should not be invoked automatically?
2- Why it appears only when all the tests run (and not unitary)

**Edit:** I've created a branch with a test which raises the cast issue
https://github.com/danielpetisme/golo-lang/tree/fix-268

**Comment by jponge** [comment]
Where do you get that error? I can build Golo just fine :-)

Java 8u60

**Comment by jponge** [comment]
`StringBuilder::append` is overloaded, so it is likely that the first invocation binds the callsite to `append(Integer)`, and hence it's why `:toString()` fixes the problem.

Golo (like at least Jython) does not deal with overloaded Java methods very well. It's a bit tricky to mount efficient guards over an array of parameters, but we can investigate a way to detect overloaded methods, and mount special (albeit likely slower) method handle chains to deal with them.

**Comment by danielpetisme** [comment]
What I propose:
1- Make a PR adding a `toString()` (or mergind the `Tuple` and `List` algos). 
2- Create the appropriate test case for this issue

**Comment by jponge** [comment]
Go for it!

**Comment by jponge** [comment]
Fixed by #269 and #285 it seems.
